docs/proposals/PROPOSAL_D.md

```diff
@@ -0,0 +1,519 @@
+        "target": "hardware.gpu.driver.version",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.architecture",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.version",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu",
+        "target": "hardware.gpu.cuda",
+        "relation": "contains"
+      },
+      {
+        "source": "hardware.gpu.cuda",
+        "target": "hardware.gpu.cuda.version",
+        "relation": "contains"
+      },
+      {
+        "source": "communication",
+        "target": "communication.framework",
+        "relation": "contains"
+      }
+    ],
+    "metadata": {
+      "version": "0.0.0",
+      "source": "https://github.com/supercontainers/compspec"
+    }
+  }
+}
+```
+
+## Compatibility Artifact
+
+The compatibility artifact would include a subset of metadata attributes that are known to a schema, along with versions and links to the schemas used.
+
+### Example Compatibility Artifact
+
+From the example above we might derive the following compatibility artifact.
+
+```json
+{
+  "version": "0.0.0",
+  "kind": "CompatibilitySpec",
+  "metadata": {
+    "name": "lammps-prototype",
+    "jsonSchema": "https://raw.githubusercontent.com/supercontainers/compspec/main/supercontainers/compspec.json"
+  },
+  "compatibilities": [
+    {
+      "name": "org.supercontainers",
+      "version": "0.0.0",
+      "annotations": {
+        "mpi.implementation": "mpich",
+        "mpi.version": "4.1.1",
+        "os.name": "Ubuntu 22.04.3 LTS",
+        "os.release": "22.04.3",
+        "os.vendor": "ubuntu",
+        "os.version": "22.04",
+        "hardware.gpu.available": "yes"
+      }
+    }
+  ]
+}
+```
+
+Note that we could easily add another compatibility interest group namespace, e.g, [archspec.io](https://github.com/supercontainers/compspec/blob/main/archspec/compspec.json):
+
+```json
+{
+  "version": "0.0.0",
+  "kind": "CompatibilitySpec",
+  "metadata": {
+    "name": "lammps-prototype"
+  },
+  "compatibilities": [
+    {
+      "name": "io.archspec",
+      "version": "0.0.0",
+      "schema": "https://raw.githubusercontent.com/supercontainers/compspec/main/archspec/compspec.json",
+      "annotations": {
+        "cpu.model": "13th Gen Intel(R) Core(TM) i5-1335U",
+        "cpu.target": "amd64",
+        "cpu.vendor": "GenuineIntel"
+      }
+    },
+    {
+      "name": "org.supercontainers",
+      "version": "0.0.0",
+      "schema": "https://raw.githubusercontent.com/supercontainers/compspec/main/supercontainers/compspec.json",
+      "annotations": {
+        "mpi.implementation": "mpich",
+        "mpi.version": "4.1.1",
+        "os.name": "Ubuntu 22.04.3 LTS",
+        "os.release": "22.04.3",
+        "os.vendor": "ubuntu",
+        "os.version": "22.04",
+        "hardware.gpu.available": "yes"
+      }
+    }
```

Comment:
if first io.archspec is about CPU, and the second one is about mpi and GPU, how to express the compatibility requirements: 1) cpu GenuineIntel amd64 + mpi v1.1 + nvidida GPU + (nvidia infiniband or arista infiband), 2) cpu GenuineAMD amd64 + mpiv1.2 + AMD GPU + arista infiniband)

Reply:
That is also up to the tool. The relationships between things are defined by the upper level schema, if that is desired, but it doesn't have to be used. This maps easily to proposal A's flat list of annotations by simply providing the key value pairs, just in a different format. For the AND case that you mentioned and a tool that requires that, you would simply look at those particular attributes and ensure they match what you need. The use case is easily met.

Reply:
I would argue that the use case is easily met.

If you define a relationship in a upper level schema, users will not be able to express their own relationships dictated by their containerized applications. They will be restricted to relations built by the schema creators (the interest group).

> For the AND case that you mentioned and a tool that requires that, you would simply look at those particular attributes and ensure they match what you need.

As an image consumer who wants to know the image requirements, I would need to look at the attributes and find a combination that will work on my host. It would be good if the author of the image could freerly specify the compatibility with its relations, so that consumers get immediate information whether the container can run on the host or not, without looking at the attributes and trying to guess what will work or not.

This is related to the comment I made earlier
> In our both proposals we expressed different requirements about where to build graph representation. I proposed on the spec level that is attached to the artifact. You on the schema level. I think that's a good point for the spec/schema discussion.

Relations on schema and spec levels solve different things. We are going to have a discussion about both approaches anyway.

Again, the comment doesn't require any changes or block the proposal.

Reply:
> f you define a relationship in a upper level schema, users will not be able to express their own relationships dictated by their containerized applications. They will be restricted to relations built by the schema creators (the interest group).

We've discussed this before, I don't think it belongs hard coded in the spec. https://github.com/opencontainers/wg-image-compatibility/pull/9#discussion_r1487725516. I don't understand why the same question is being asked again. I'll reiterate I'm open to suggestions and changes in the design to accommodate this use case, despite not liking the move toward more complexity.